tests/unit/library/test_json_stringformat.py

```diff
@@ -0,0 +1,941 @@
+            '"99c17cbb-656f-564a-940f-1a4568f03487"',  # valid version 5
+            '"99c17cbb-656f-664a-940f-1a4568f03487"',  # hypothetical version 6
+            '"99c17cbb-656f-f64a-940f-1a4568f03487"',  # hypothetical version 15
+        ],
+    )
+    def test_good(self, target_str):
+        schema_obj = json.loads(self.schema)
+        target_obj = json.loads(target_str)
+        generate_and_check(target_obj, schema_obj)
+
+    @pytest.mark.parametrize(
+        "bad_str",
+        [
+            '"2eb8aa08-aa98-11ea-b4aa-73b441d1638"',  # wrong length
+            '"2eb8aa08-aa98-11ea-73b441d16380"',  # missing section
+            '"2eb8aa08-aa98-11ea-b4ga-73b441d16380"',  # bad characters (not hex)
+            '"2eb8aa08aa9811eab4aa73b441d16380"',  # no dashes
+            '"2eb8aa08aa98-11ea-b4aa73b441d16380"',  # too few dashes
+            '"2eb8-aa08-aa98-11ea-b4aa73b44-1d16380"',  # too many dashes
+            '"2eb8aa08aa9811eab4aa73b441d16380----"',  # dashes in the wrong spot
+        ],
+    )
+    def test_bad(self, bad_str):
+        schema_obj = json.loads(self.schema)
+        check_match_failure(bad_string=bad_str, schema_obj=schema_obj, compact=True)
+
+
+class TestEmail:
+    schema = '{"$schema":"https://json-schema.org/draft/2020-12/schema","format":"email"}'
+
+    @pytest.mark.parametrize(
+        "target_str",
+        [
+            "12",  # all string formats ignore integers
+            "13.7",  # all string formats ignore floats
+            "{}",  # all string formats ignore objects
+            "[]",  # all string formats ignore arrays
+            "false",  # all string formats ignore booleans
+            "null",  # all string formats ignore nulls
+            '"joe.bloggs@example.com"',  # a valid e-mail address
+            '"te~st@example.com"',  # tilde in local part is valid
+            '"~test@example.com"',  # tilde before local part is valid
+            '"test~@example.com"',  # tilde after local part is valid
+            pytest.param('"\\"joe bloggs\\"@example.com"', marks=pytest.mark.xfail(reason="Quoted strings not yet implemented in local part")),  # a quoted string with a space in the local part is valid
+            pytest.param('"\\"joe..bloggs\\"@example.com"', marks=pytest.mark.xfail(reason="Quoted strings not yet implemented in local part")),  # a quoted string with a double dot in the local part is valid
+            pytest.param('"\\"joe@bloggs\\"@example.com"', marks=pytest.mark.xfail(reason="Quoted strings not yet implemented in local part")),  # a quoted string with a @ in the local part is valid
+            '"joe.bloggs@[127.0.0.1]"',  # an IPv4-address-literal after the @ is valid
+            pytest.param('"joe.bloggs@[IPv6:::1]"', marks=pytest.mark.xfail(reason="IPv6 is hard")),  # an IPv6-address-literal after the @ is valid
+            '"te.s.t@example.com"',  # two separated dots inside local part are valid
```

Comment:
Small extra one (since things not supporting it bug me): can you include a local part with a `+` in the middle, like GMail supports?

Reply:
Can do :)